src/MongoDB/Session.c

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright 2014-2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <php.h>
+#include <Zend/zend_interfaces.h>
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+#include "php_bson.h"
+
+zend_class_entry *php_phongo_session_ce;
+
+/* {{{ proto void MongoDB\Driver\Session::advanceClusterTime(array|object $clusterTime)
+   Advances the cluster time for this Session */
+static PHP_METHOD(Session, advanceClusterTime)
+{
+	php_phongo_session_t *intern;
+	zval                 *zcluster_time;
+	bson_t                cluster_time = BSON_INITIALIZER;
+	SUPPRESS_UNUSED_WARNING(return_value_ptr) SUPPRESS_UNUSED_WARNING(return_value_used)
+
+
+	intern = Z_SESSION_OBJ_P(getThis());
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "A", &zcluster_time) == FAILURE) {
```

Comment:
I presume we shouldn't need to verify the structure of this cluster time object?

Reply:
Good question. libmongoc has a private [`_mongoc_parse_cluster_time()`](https://github.com/mongodb/mongo-c-driver/blob/6fa2c35119672c8ce7c0ac9a926fc33ecc3b7ae1/src/mongoc/mongoc-client-session.c#L146) function used to validate its structure. The driver only cares about the timestamp and doesn't validate the signature (only used server-side AFAIK).

[`mongoc_client_session_advance_cluster_time()`](https://github.com/mongodb/mongo-c-driver/blob/6fa2c35119672c8ce7c0ac9a926fc33ecc3b7ae1/src/mongoc/mongoc-client-session.c#L351) only assigns it if it parses successfully, and the function is void so we don't even get an error if it's malformed. I'm curious if that was intentional. Perhaps @ajdavis can chime in.

On a related note, [the spec](https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#gossipping-the-cluster-time) doesn't refer to cluster time as an "opaque value" like we do for session and transaction IDs in retryable writes. Therefore, users can certainly access it and poke around with it. The spec also doesn't say anything about reporting errors if users try to advance an invalid value (see [here](https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#advanceclustertime)).

Reply:
Right, we only care that it has a timestamp, we don't know or care about other fields. I didn't think it worthwhile to implement elaborate error handling. Applications should simply save a cluster time from a session, then set it on another session, without modification. If you make a mistake there's an error log.